## Import Libraries
We are importing the following libraries:
- **pandas**: Used for data manipulation and analysis, especially with dataframes.
- **numpy**: Provides tools for numerical computations, including arrays and mathematical operations.
- **string**: Contains a collection of string constants and functions useful for string manipulation.


In [1]:
import pandas as pd
import numpy as np
import string


## Data Cleaning and Tokenization

1. **`get_and_clean_data`**: 
   - Reads data from a CSV file.
   - Cleans the data (removes NaN, strips punctuation except `#` and `+`, converts to lowercase, normalizes whitespace).
   - Removes duplicate rows.

2. **`simple_tokenize`**: 
   - Tokenizes text in `job_title` and `job_description` into individual words.


In [2]:
# Define the columns to filter from the dataset
filters = ['job_title', 'job_description']

# Function to load and clean the dataset
def get_and_clean_data():
    # Load dataset from a CSV file
    data = pd.read_csv('data/software_developer_united_states_1971_20191023_1.csv')
    
    # Keep only the specified columns (filters)
    filtered_data = data[filters]
    
    # Iterate through each column in the filters
    for column in filters:
        # Replace NaN values with an empty string
        filtered_data[column] = filtered_data[column].fillna('')
        
        # Remove punctuation except '#' and '+'
        filtered_data[column] = filtered_data[column].apply(
            lambda s: s.translate(
                str.maketrans('', '', string.punctuation.replace('#', '').replace('+', ''))
            )
        )
        
        # Convert text to lowercase
        filtered_data[column] = filtered_data[column].apply(lambda s: s.lower())
        
        # Normalize whitespace characters
        filtered_data[column] = filtered_data[column].apply(
            lambda s: s.translate(str.maketrans(string.whitespace, ' ' * len(string.whitespace), ''))
        )
    
    # Remove duplicate rows
    filtered_data = filtered_data.drop_duplicates()
    
    # Return the cleaned dataset
    return filtered_data

# Function to tokenize text in each column of the dataset
def simple_tokenize(data):
    for column in filters:
        # Split text into tokens and strip whitespace from each token
        data[column] = data[column].apply(lambda s: [x.strip() for x in s.split()])
    return data


## Display Cleaned and Tokenized Data
- Cleans and tokenizes the dataset using the defined functions.
- Displays the first 10 rows in markdown table format for easy visualization.


In [3]:
# Clean and tokenize data
data = get_and_clean_data()
data = simple_tokenize(data)

# Display the first 10 rows
print(data.head(10))


C:\Users\Admin\AppData\Local\Temp\ipykernel_30056\2774603895.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data[column] = filtered_data[column].fillna('')
C:\Users\Admin\AppData\Local\Temp\ipykernel_30056\2774603895.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  filtered_data[column] = filtered_data[column].apply(
C:\Users\Admin\AppData\Local\Temp\ipykernel_30056\2774603895.py:25: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc

                                     job_title  \
0                    [sr, software, developer]   
1              [c#, lead, software, developer]   
2                [senior, software, developer]   
3                [senior, software, developer]   
4              [c#, lead, software, developer]   
5                     [software, developer, 3]   
6                   [c++, software, developer]   
7  [c++, software, developer, local, w2, only]   
8                   [software, developer, net]   
9                   [net, software, developer]   

                                     job_description  
0  [the, chosen, sr, software, developer, will, b...  
1  [position, c#, lead, software, developer, loca...  
2  [senior, software, developer, hoboken, nj, sta...  
3  [our, client, a, multinational, publishing, an...  
4  [position, c#, lead, software, developer, loca...  
5  [responsibilities, kforce, has, a, client, see...  
6  [apply, by, emaildirect, application, at, udit...  
7  [talen

## Filter for Junior Roles
- Filters the dataset for job titles containing "junior" or "jr".
- Displays the total count and the first 10 rows of filtered results in markdown table format.

### Q1: Identify database and programming language proficiencies essential for a junior-level (2pts, should be achievable by everyone)

In [4]:
# Filter rows where 'junior' or 'jr' is mentioned in the job title
junior_roles = data[data['job_title'].apply(lambda tokens: 'junior' in tokens or 'jr' in tokens)]

# Display the count of filtered jobs and the first 10 rows
print(f'Job counts = {len(junior_roles)}')
print(junior_roles.head(10))


Job counts = 513
                                             job_title  \
20                       [junior, software, developer]   
40                       [junior, software, developer]   
46                           [jr, software, developer]   
52                       [junior, software, developer]   
68   [junior, to, mid, level, software, developer, ...   
90   [junior, level, software, developer, 14, years...   
163  [junior, software, developer, engineer, in, te...   
178                  [junior, c#, software, developer]   
183                      [junior, software, developer]   
197                      [junior, software, developer]   

                                       job_description  
20   [about, harmer, consultants, inc, harmer, cons...  
40   [responsibilities, kforce, has, a, client, see...  
46   [jr, software, developer, about, acet, adams, ...  
52   [responsibilities, kforce, has, a, client, see...  
68   [general, dynamics, information, technology, a...  
90

## Programming Languages Analysis
- Counts mentions of specific programming languages in the `job_description` column.
- Sorts the languages by frequency in descending order.
- Displays the ranked list of programming languages with their counts.

### This one we compare the languages and count the frequency

In [5]:
# List of programming languages to count in job descriptions
program_languages = ['c', 'c#', 'c++', 'java', 'python', 'kotlin', 'swift', 'rust', 'ruby', 'scala', 'julia', 'lua']

# Initialize a dictionary to store counts for each language
languages_counter = {}

for pl in program_languages:
    # Count occurrences of each language in job descriptions
    count = junior_roles['job_description'].apply(lambda s: pl in s).sum()
    languages_counter[pl] = count

# Sort the programming languages by count in descending order
languages_counter = sorted(languages_counter.items(), key=lambda x: x[1], reverse=True)

# Print the rankings
print("For programming Languages: ")
for lang, count in languages_counter:
    print(f' - {lang}: {count}')


For programming Languages: 
 - java: 270
 - c#: 194
 - c++: 129
 - python: 88
 - c: 75
 - ruby: 73
 - swift: 10
 - scala: 5
 - kotlin: 0
 - rust: 0
 - julia: 0
 - lua: 0


## Database and Cloud Service Analysis
- Counts mentions of specific databases and cloud services in the `job_description` column.
- Sorts the results by frequency in descending order.
- Displays the ranked list of databases and cloud services with their counts.

### And this one is for data base

In [6]:
# List of databases and cloud services to analyze
databases = ['mysql', 'postgresql', 'mssql', 'oracle', 'sqlite', 
             'mongodb', 'cassandra', 'redis', 'dynamodb', 'firebase', 
             'neo4j', 'cloudsql', 'aws', 'azure', 'bigquery']

# Initialize a dictionary to store counts for each database
databases_counter = {}

for db in databases:
    # Count occurrences of each database in job descriptions
    count = junior_roles['job_description'].apply(lambda s: db in s).sum()
    databases_counter[db] = count

# Sort the databases by count in descending order
databases_counter = sorted(databases_counter.items(), key=lambda x: x[1], reverse=True)

# Print the rankings
print("For databases: ")
for db, count in databases_counter:
    print(f' - {db}: {count}')


For databases: 
 - oracle: 101
 - mysql: 86
 - aws: 34
 - azure: 19
 - mssql: 11
 - postgresql: 10
 - mongodb: 8
 - sqlite: 3
 - cassandra: 3
 - redis: 3
 - dynamodb: 2
 - neo4j: 1
 - firebase: 0
 - cloudsql: 0
 - bigquery: 0


### So the answer of question 1 is top languages is java and the database is oracle

---

#### Q2: For a long-term skill development plan, determine an additional programming language to complement your initial choice for a first job. (2 pts – achievable with extra effort)
Hints:
- Analyze effective pairings between your selected language and one another programming language at time.
- Focus only on senior-level role.


In [7]:
# Filter rows where 'senior' or 'sr' is mentioned in the job title
senior_roles = data[data['job_title'].apply(lambda tokens: 'senior' in tokens or 'sr' in tokens)]

# Display the count of filtered jobs and the first 10 rows
print(f'Job counts = {len(senior_roles)}')
print(senior_roles.head(10))


Job counts = 1583
                                     job_title  \
0                    [sr, software, developer]   
2                [senior, software, developer]   
3                [senior, software, developer]   
10         [senior, java, software, developer]   
15               [senior, software, developer]   
23               [software, developer, senior]   
25               [senior, software, developer]   
26  [senior, full, stack, software, developer]   
27          [software, developer, sr, advisor]   
35                   [sr, software, developer]   

                                      job_description  
0   [the, chosen, sr, software, developer, will, b...  
2   [senior, software, developer, hoboken, nj, sta...  
3   [our, client, a, multinational, publishing, an...  
10  [job, summary, biotelemetry, provides, realtim...  
15  [minimum, required, skills, net, c#, visual, s...  
23  [read, what, people, are, saying, about, worki...  
25  [the, denzel, group, has, been, cho

In [8]:
# Filter senior roles by the most mentioned programming language (from the previous ranking)
top_senior_roles = senior_roles[senior_roles['job_description'].apply(lambda tokens: languages_counter[0][0] in tokens)]

# Display the count of filtered senior roles and the first 10 rows
print(f'Job counts = {len(top_senior_roles)}, filtered by {languages_counter[0][0]}')
print(top_senior_roles.head(10))


Job counts = 706, filtered by java
                                            job_title  \
10                [senior, java, software, developer]   
23                      [software, developer, senior]   
27                 [software, developer, sr, advisor]   
38                    [sr, lead, software, developer]   
44  [senior, java, software, developerengineer, im...   
51        [sr, software, developer, java, cloud, api]   
57                    [sr, java, software, developer]   
67                 [sr, steppim, software, developer]   
81  [software, developer, analyst, ii, or, senior,...   
93    [senior, frontend, software, developer, remote]   

                                      job_description  
10  [job, summary, biotelemetry, provides, realtim...  
23  [read, what, people, are, saying, about, worki...  
27  [read, what, people, are, saying, about, worki...  
38  [employer, nyse, ctl, is, a, global, communica...  
44  [position, overview, java, software, developer...  
5

#### Count the filtered data again

In [9]:
# Remove the top programming language (most mentioned) from the list
program_languages.remove(languages_counter[0][0])

# Initialize a dictionary to store co-occurrence counts for remaining languages
co_occurrence_counts = {}

for pl in program_languages:
    # Count occurrences of each remaining language in the job descriptions of top senior roles
    count = top_senior_roles['job_description'].apply(lambda tokens: pl in tokens).sum()
    co_occurrence_counts[pl] = count

# Sort the co-occurrence counts in descending order
co_occurrence_counts = sorted(co_occurrence_counts.items(), key=lambda x: x[1], reverse=True)

# Print the co-occurrence counts for the remaining languages
print(f'Co-occurrence of programming languages with {languages_counter[0][0]}:')
for lang, count in co_occurrence_counts:
    print(f" - {lang}: {count}")

# Identify the most co-occurring programming language
additional_programming_language = co_occurrence_counts[0][0]

Co-occurrence of programming languages with java:
 - c#: 217
 - python: 149
 - c++: 98
 - c: 73
 - ruby: 45
 - scala: 25
 - swift: 16
 - kotlin: 11
 - lua: 1
 - rust: 0
 - julia: 0


### Answer of question 2 is c# (217)

---

### Question3:Design and execute an experiment that determines which database proefficiency is best suited for a senior software developer role. (1pt – challenging assignment)

- Apply Bayes' theorem and code the formula to support your analysis.<br>
- Use Monte Carlo simulations with 10,000 iterations to demonstrate your conclusions.<br>
- Hint: P(senior software developer) is around 7% and you already know each P(skill) already. You will need to assume and code their distributions.

### Define Bayes Theorem

#### Probability Calculations
1. **P(senior)**: Calculates the probability of a job being a senior role in the dataset.
2. **P(skill)**: Calculates the probability of each skill (database) appearing in all job descriptions.
3. **P(skill | senior)**: Calculates the probability of each skill (database) appearing in senior roles.

Displays the probabilities for:
- **P(senior)**
- **P(skill)**
- **P(skill | senior)**


In [10]:
# P(senior)
P_senior = len(senior_roles) / len(data)

# Calculate P(skill) and P(skill | senior)
P_skill = {}
P_skill_given_senior = {}

for db in databases:
    # P(skill): All jobs mentioning the skill
    P_skill[db] = data['job_description'].apply(lambda s: db in s).sum() / len(data)
    
    # P(skill | senior): Senior jobs mentioning the skill
    P_skill_given_senior[db] = senior_roles['job_description'].apply(lambda s: db in s).sum() / len(senior_roles)

print("P(senior):", P_senior)
print("P(skill):", P_skill)
print("P(skill | senior):", P_skill_given_senior)

P(senior): 0.2075793338578547
P(skill): {'mysql': 0.08785733018620509, 'postgresql': 0.03448728035667453, 'mssql': 0.014424337791765015, 'oracle': 0.18345135064253867, 'sqlite': 0.0036716496197220037, 'mongodb': 0.03920797272488854, 'cassandra': 0.01888276947285602, 'redis': 0.014030946761080514, 'dynamodb': 0.011670600576973512, 'firebase': 0.0013113034356150013, 'neo4j': 0.0036716496197220037, 'cloudsql': 0.00026226068712300026, 'aws': 0.13335955940204564, 'azure': 0.07802255441909257, 'bigquery': 0.0005245213742460005}
P(skill | senior): {'mysql': 0.07770056854074542, 'postgresql': 0.028427037271004423, 'mssql': 0.007580543272267846, 'oracle': 0.1945672773215414, 'sqlite': 0.006317119393556538, 'mongodb': 0.04611497157296273, 'cassandra': 0.02021478205938092, 'redis': 0.02147820593809223, 'dynamodb': 0.013265950726468731, 'firebase': 0.0012634238787113076, 'neo4j': 0.005685407454200884, 'cloudsql': 0.0, 'aws': 0.20025268477574226, 'azure': 0.09665192672141504, 'bigquery': 0.0}


#### Calulate with bayes theorem

In [11]:
P_senior_given_skill = {}

for db in databases:
    if P_skill[db] > 0:  # Avoid division by zero
        P_senior_given_skill[db] = (P_skill_given_senior[db] * P_senior) / P_skill[db]
    else:
        P_senior_given_skill[db] = 0

# Display results
sorted_probs = sorted(P_senior_given_skill.items(), key=lambda x: x[1], reverse=True)
print("P(senior | skill) Ranking:")
for db, prob in sorted_probs:
    print(f" - {db}: {prob:.4f}")

P(senior | skill) Ranking:
 - sqlite: 0.3571
 - neo4j: 0.3214
 - redis: 0.3178
 - aws: 0.3117
 - azure: 0.2571
 - mongodb: 0.2441
 - dynamodb: 0.2360
 - cassandra: 0.2222
 - oracle: 0.2202
 - firebase: 0.2000
 - mysql: 0.1836
 - postgresql: 0.1711
 - mssql: 0.1091
 - cloudsql: 0.0000
 - bigquery: 0.0000


### Monte Carlo Simulation for Senior Role Probabilities
- Runs **Monte Carlo simulations** (10,000 iterations) to estimate the probability of a senior role given a skill.
- Simulates the occurrence of a skill and a senior role using binomial distributions based on **P(skill)** and **P(senior)**.
- Applies **Bayes' Theorem** to compute the conditional probability **P(senior | skill)**.
- Displays the average results for each database and ranks them by the highest probability.


In [12]:
# Number of Monte Carlo iterations
iterations = 10000

# Initialize dictionary to store results for each database
monte_carlo_results = {db: [] for db in databases}

# Run Monte Carlo simulations
for db in databases:
    for _ in range(iterations):
        # Simulate skill occurrence based on P(skill) using binomial distribution
        skill_occurrence = np.random.binomial(1, P_skill[db])
        
        # Simulate senior role occurrence based on P(senior) using binomial distribution
        senior_occurrence = np.random.binomial(1, P_senior)
        
        if skill_occurrence and senior_occurrence:
            # Apply Bayes' Theorem to calculate P(senior | skill)
            P_senior_given_skill = (P_skill_given_senior[db] * P_senior) / P_skill[db]
            monte_carlo_results[db].append(P_senior_given_skill)

# Calculate average results from the Monte Carlo simulations
average_results = {db: np.mean(monte_carlo_results[db]) for db in databases}

# Sort results and display the rankings
sorted_monte_carlo = sorted(average_results.items(), key=lambda x: x[1], reverse=True)
print("Monte Carlo Simulation Rankings:")
for db, prob in sorted_monte_carlo:
    print(f" - {db}: {prob:.4f}")

# Identify the top-ranked database
top_sim = sorted_monte_carlo[0]


Monte Carlo Simulation Rankings:
 - sqlite: 0.3571
 - neo4j: 0.3214
 - redis: 0.3178
 - cloudsql: nan
 - aws: 0.3117
 - azure: 0.2571
 - mongodb: 0.2441
 - dynamodb: 0.2360
 - cassandra: 0.2222
 - oracle: 0.2202
 - firebase: 0.2000
 - mysql: 0.1836
 - postgresql: 0.1711
 - mssql: 0.1091
 - bigquery: 0.0000


C:\Users\Admin\anaconda3\envs\new\Lib\site-packages\numpy\core\fromnumeric.py:3504: RuntimeWarning: Mean of empty slice.
  return _methods._mean(a, axis=axis, dtype=dtype,
C:\Users\Admin\anaconda3\envs\new\Lib\site-packages\numpy\core\_methods.py:129: RuntimeWarning: invalid value encountered in scalar divide
  ret = ret.dtype.type(ret / rcount)


### **Answer question 3:**  sqlite the top from simulation has the highest posterior probability